toolchain/check/custom_witness.cpp

```diff
@@ -132,4 +208,95 @@ auto GetCoreInterface(Context& context, SemIR::InterfaceId interface_id)
+  auto destroy_id = LookupNameInCore(context, loc_id, CoreIdentifier::Destroy);
+  auto destroy_facet_type =
+      context.insts().TryGetAs<SemIR::FacetType>(destroy_id);
+  if (!destroy_facet_type) {
+    return SemIR::InterfaceId::None;
+  }
+  const auto& destroy_facet_type_info =
+      context.facet_types().Get(destroy_facet_type->facet_type_id);
+  auto destroy_extend = destroy_facet_type_info.TryAsSingleExtend();
+  if (!destroy_extend ||
+      !std::holds_alternative<SemIR::SpecificInterface>(*destroy_extend)) {
+    return SemIR::InterfaceId::None;
+  }
+
+  return std::get<SemIR::SpecificInterface>(*destroy_extend).interface_id;
+}
+
+// Returns true if the `Self` should impl `Destroy`.
+static auto TypeCanDestroy(Context& context, SemIR::LocId loc_id,
+                           SemIR::ConstantId query_self_const_id) -> bool {
+  auto inst = context.insts().Get(context.constant_values().GetInstId(
+      GetCanonicalFacetOrTypeValue(context, query_self_const_id)));
+
+  // For facet values, look if the FacetType provides the same.
+  if (auto facet_type =
+          context.types().TryGetAs<SemIR::FacetType>(inst.type_id())) {
+    auto destroy_interface_id = GetDestroyInterface(context, loc_id);
+    if (!destroy_interface_id.has_value()) {
+      return false;
+    }
+    const auto& info = context.facet_types().Get(facet_type->facet_type_id);
+    for (auto interface : info.extend_constraints) {
+      if (interface.interface_id == destroy_interface_id) {
+        return true;
+      }
+    }
+  }
+
+  CARBON_KIND_SWITCH(inst) {
+    case CARBON_KIND(SemIR::ClassType class_type): {
+      auto class_info = context.classes().Get(class_type.class_id);
+      // Incomplete and abstract classes can't be destroyed.
+      // TODO: Return false if the object repr doesn't impl `Destroy`.
+      // TODO: This should probably be skipped for all C++ types, but currently
+      // must handle those for trivial destruction.
+      return class_info.is_complete() &&
+             class_info.inheritance_kind !=
+                 SemIR::Class::InheritanceKind::Abstract;
+    }
+    case SemIR::ArrayType::Kind:
+    case SemIR::ConstType::Kind:
+    case SemIR::MaybeUnformedType::Kind:
+    case SemIR::PartialType::Kind:
+    case SemIR::StructType::Kind:
+    case SemIR::TupleType::Kind:
+      // TODO: Return false for types that indirectly reference a type that
+      // doesn't impl `Destroy`.
+      return true;
+    case SemIR::BoolType::Kind:
+    case SemIR::PointerType::Kind:
+      // Trivially destructible.
+      return true;
+    default:
+      return false;
+  }
+}
+
+auto LookupDestroyCustomWitness(
+    Context& context, SemIR::LocId loc_id,
+    SemIR::ConstantId query_self_const_id,
+    SemIR::SpecificInterfaceId query_specific_interface_id,
+    SemIR::SpecificInterface query_specific_interface) -> SemIR::InstId {
+  // If we're looking up `Destroy`, first try builtin support without looking at
+  // `impl`s.
```

Comment:
Does this comment match the if/early out? It looks like we're checking if we should try to destroy at all. I don't see any builtin support relationship in TypeCanDestroy?

Reply:
Removed, this comment is obsolete and I copied around too much.